### PR TITLE
Ruby Group Capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by JuanDiegoGuerra**
### Summary
This PR represents the culmination of our collaborative efforts to build a comprehensive catalog management application using Ruby. Our team divided tasks and created distinct classes and functionalities as specified in the project requirements.
> Pair Programming ( @Rysth @hafiz1379 @JuanDiegoGuerra ) 🧙🏼‍♂️👨🏼‍🚀👨🏼‍💻
### Documentation
- The content of the video should include:
  - `1. `A description of the project. `2.` A demo of the project features. `3.` Highlight some interesting piece of code or something you built that you are very proud of. `4.`You all should appear in the video and talk about the project.
- Video Link: [Ruby Group Project Presentation](https://drive.google.com/file/d/1rHeHR-GSEvMwF4JfDjMF8RqTOvctFEyk/view?usp=sharing)
### Changes Made
**Team Member #1 - Books and Labels**
- Created the `Book` class in a separate `.rb` file.
- Designed and implemented the `Label` class with an association to the `Item` class in a separate `.rb` file.
- Ensured that properties and methods from the diagram were correctly defined and set up in the constructor methods of both classes.
- Implemented the `add_item` method in the `Label` class to add items to the collection of labels.
- Added the `can_be_archived?` method in the `Book` class to determine if a book can be archived based on cover condition and publication date.
- Included comprehensive unit tests to validate the correctness of the implemented methods.

**Team Member #2 - Music Albums and Genres**
- Created the `MusicAlbum` class in a separate `.rb` file.
- Designed and implemented the `Genre` class with an association to the `Item` class in a separate `.rb` file.
- Ensured that properties and methods from the diagram were correctly defined and set up in the constructor methods of both classes.
- Implemented the `add_item` method in the `Genre` class to add items to the collection of genres.
- Added the `can_be_archived?` method in the `MusicAlbum` class to determine if a music album can be archived based on Spotify availability and parent's archiving criteria.
- Included comprehensive unit tests to validate the correctness of the implemented methods.

**Team Member #3 - Games and Authors**
- Created the `Game` class in a separate `.rb` file.
- Designed and implemented the `Author` class with an association to the `Item` class in a separate `.rb` file.
- Ensured that properties and methods from the diagram were correctly defined and set up in the constructor methods of both classes.
- Implemented the `add_item` method in the `Author` class to add items to the collection of authors.
- Added the `can_be_archived?` method in the `Game` class to determine if a game can be archived based on last played date and parent's archiving criteria.
- Included comprehensive unit tests to validate the correctness of the implemented methods.

**User Interface - `main.rb`**
- Created a `main.rb` file to serve as the console app entry-point.
- Designed a user-friendly interface that presents a list of options for managing catalogs and items.
- Implemented interactive features that allow users to choose options, input parameters as needed, and quit the application.
- Provided functionality to list books, music albums, games, genres, labels, and authors, as well as add new items.
- Ensured data persistence by saving collections in .json files.

**Database - `schema.sql`**
- Created a `schema.sql` file with tables analogical to the structure of the classes.
- Designed tables for books, music albums, games, genres, labels, and authors, including columns representing properties and associations from the parent `Item` class.
- Ensured that the schema.sql file can be used to store data in the future.

### Reviewer's Notes
We invite the code reviewer to thoroughly examine our collaborative project. Please review the classes, methods, unit tests, user interface, and database schema to ensure they align with the original project requirements. Your feedback on code quality, correctness, and data persistence is highly appreciated. Thank you for taking the time to review this PR.
This PR reflects our collective effort and collaboration in building a robust and versatile catalog management system.

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        id: book_data['id'],
+        title: book_data['title'], # Asegúrate de que esto esté en tus datos JSON
+        publisher: book_data['publisher'],
+        cover_state: book_data['cover_state'],
+        publish_date: book_data['publish_date'],
+        label: find_label_by_title(book_data['label'])
+      )
+      @books << book
+    end
+  end
```
Comment (codepantha) [inline app.rb]: - [x] I noticed that the app crashes if the `.json` files are non-existent of if they are empty. Please implement a fix that ensures the app doesn't crash the aforementioned cases 👍 
image

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights 🎯 
✔️ Descriptive PR
✔️ Clear demo video 📌 
✔️ Wrote unit tests 📌 
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+require_relative 'app'
+app = App.new
+app.run
```
Comment (tresorsawasawa) [inline main.rb:3]: - [x] As mentioned by the previous reviewer and for the project requirements, kindly ensure that your app is not crashing while testing it. I recommend you fix all errors as now your app is still showing the below error while trying to test it. After showing all the options, no-one can work. It is only shows the basic options, and any one can work(i.e. your app can only run basic options)
image

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUEST CHANGES :recycle::recycle: 

Hi @JuanDiegoGuerra, @Rysth, and @hafiz1379 
Great work on making the changes requested by the previous reviewer  👏👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost-there

## To highlight :champagne: 
- Most of your associations are correct :heavy_check_mark: 
- Good PR description :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by AbelG101 — APPROVED** [review]
# Hi @JuanDiegoGuerra, @Rysth & @hafiz1379 👋, 

**Your project is complete!  There is nothing else to say other than... it's time to merge it. 
Congratulations! 🎉 🥳🥇**

## Highlights
- UI is looking awesome ✔
- CLI works & transitions perfectly ✔
- Data is properly preserved & retrieved in/from JSON files ✔
- Case where file doesn't exist is properly handled ✔
- Correct use of GitHub Flow ✔
-  Well-structured files ✔
- All linter checks are passing ✔
- Descriptive PR title with a great summary ✔

## Optional suggestions
- N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by AbelG101** [comment]
Hello guys, for some reason the dashboard didn't approve I'll have to reapprove it for you. Thank you!

**Comment by hafiz1379** [comment]
Thank you @AbelG101

**Pull request merged**